[MIG][10.0] auth_session_timeout

lasley opened the pull request:
WIP - I need to add test coverage for `res.users`
- Bump versions
- Installable to True
- Add Usage section to ReadMe w/ Runbot link
- `_crypt_context` now directly exposes the `CryptContext`
- Change all instances of openerp to odoo
- Add test coverage to IrConfigParameter

lasley committed:
[MIG] auth_session_timeout: Migrate to v10
* Bump versions
* Installable to True
* Add Usage section to ReadMe w/ Runbot link
* `_crypt_context` now directly exposes the `CryptContext`
* Change all instances of openerp to odoo
* Add test coverage to IrConfigParameter
* Add test coverage for res.users
* Remove db from `get_session_parameters` method call
* Remove deprecated skiparg for ormcache
* Fix tests & lint
* Switch cache to use self.cr.dbname
* Fix ormcache

lasley commented on auth_session_timeout/i18n/de.po:
```diff
@@ -1,23 +0,0 @@
-# Translation of Odoo Server.
```
@pedrobaeza - I could have sworn I saw the removal of PO files from migrations somewhere, but I now can't remember where. Am I being stupid here, or are these actually supposed to be removed?

pedrobaeza replied:
You need to remove POT files, not PO ones. Having PO files, although outdated, allow to reuse the sentences that haven't changed. Next Transifex synchronization will wipe out the non existing terms comparing with the newest POT file.

@atchuthan proposed in https://github.com/OCA/maintainer-tools/issues/232 that we should avoid the commits that updates the languages in the migration process, but I'm evaluating if this is a good idea at all.

lasley replied:
Thanks for the explanation @pedrobaeza - I think that the issue that you linked is the one that made me think this way. I'll add back 😄

lasley commented:
Coverage added in all but `res.users.check`, which was going to be a real hassle. The logic is simple enough that I don't think it will be a problem.

Ready for review

elicoidal commented on auth_session_timeout/README.rst:
```diff
@@ -13,16 +15,32 @@ Configuration
+Usage
+=====
+
+Setup the session parameters as described above.
```
You might want to move the above section in this one.

lasley replied:
I thought about it, but I'm honestly torn. They are more of a configuration instruction vs usage. Maybe instead I add `Use system like normal` in place of the note to configure using instructions?

elicoidal reviewed (approved):
:+1:

moylop260 commented:
+1

lasley commented:
Good to merge?

yajo commented on auth_session_timeout/models/ir_config_parameter.py:
```diff
@@ -13,24 +12,18 @@
     _inherit = 'ir.config_parameter'
 
+    @api.model
     @tools.ormcache(skiparg=0)
```
What should be the cache key? `skiparg` is deprecated. Better use `@tools.ormcache("self.env.db")` IMHO.

lasley replied:
OK so I'm a bit confused here & I am pretty sure this is what is causing the current error.

What are we going for with this cache? Do you maybe have an example of some other code that does something similar, or would mind giving a quick breakdown for me? My google-fu is turning up zilch here.

yajo reviewed (approved):
Just a little thing, not requirement.

nilshamerlinck commented on auth_session_timeout/models/res_users.py:
```diff
@@ -1,29 +1,25 @@
+        session = http.request.session
+        session_store = http.root.session_store
+        ConfigParam = http.request.env['ir.config_parameter']
+        delay, urls = ConfigParam.get_session_parameters(db)
```
No need for the `db` parameter anymore, right?

lasley commented:
Thanks for the reviews @Yajo & @nilshamerlinck - both points fixed.

yajo commented:
Bots still ❌

yajo commented on auth_session_timeout/models/ir_config_parameter.py:
```diff
@@ -13,24 +12,18 @@
-            cr.close()
-        return delay, urls
+    @api.model
+    @tools.ormcache('self.env.db')
```
@lasley, about https://github.com/OCA/server-tools/pull/580#discussion_r94074703 (I answer here so this does not get hidden)...

I learned the hard way to use ormcache when developing [`website_sale_wishlist`](https://github.com/OCA/e-commerce/tree/9.0/website_sale_wishlist). It turns out it's a pretty neat tool. It creates a key by the arguments you pass to it, and stores the method result in a cache attached to your current `self.env`, so further calls to the method return the cached result instead of executing the method again and again.

It's most useful when a method is CPU-expensive, is called quite often, and it returns a predictable result based on the incoming arguments/environment.

In old api, we had the `skipargs`, which was used to skip a number of arguments and use the rest to make a cache. For instance, if your method had `cr, uid, ids, key, context=None` and the result was the same no matter the `cr, uid`, you'd say `skipargs=2`.

Since v9, [we have the new notation where you give the name of the arguments that form the cache key](https://github.com/odoo/odoo/blob/9.0/openerp/tools/cache.py#L36). Since `self` is always the first argument, [you can use that one (and properties of it) to form a key](https://github.com/OCA/e-commerce/blob/9.0/website_sale_wishlist/models/product_template.py#L10-L17).

When you change your `env` (like `self.with_context(goal="I wanna be a pirate")`), cache is empty. This is logical because a different context/cr/uid usually yields different results almost everywhere. However, when you apply cache to a method, [it gets a `clear_cache()` method](https://github.com/odoo/odoo/blob/9.0/openerp/tools/cache.py#L36d) that [you can use later to clear it without having to change your environment](https://github.com/OCA/e-commerce/blob/9.0/website_sale_wishlist/models/product_wishlist.py#L93-L100).

lasley replied:
Thanks for the excellent explanation @Yajo - this totally clarifies what we are going for here & I think points me in the right direction. I'll make the update soon 😄

lasley commented:
Oops seems I forgot about this. With the excellent explanation @Yajo gave regarding the cache, I abstracted that a decent replacement for the `db` arg we had previously is now `self.cr.dbname`. 

It seems OCA/OCB just synced up with odoo/odoo though, so now we're red [for other reasons](https://odoo-community.org/groups/contributors-15/contributors-49663?mode=thread&date_begin=&date_end=). Oh the joy!

karlis-dreizis commented:
So i took this PR for a spin and I'm getting an error.
```
Traceback (most recent call last):
  File "/opt/odoo/10/odoo/odoo/addons/base/ir/ir_http.py", line 103, in _authenticate
    request.session.check_security()
  File "/opt/odoo/10/odoo/odoo/http.py", line 1056, in check_security
    security.check(self.db, self.uid, self.password)
  File "/opt/odoo/10/odoo/odoo/service/security.py", line 13, in check
    return res_users.check(db, uid, passwd)
  File "/opt/odoo/10/oca/server-tools/auth_session_timeout/models/res_users.py", line 40, in check
    cls._check_session_validity(db, uid, passwd)
  File "/opt/odoo/10/oca/server-tools/auth_session_timeout/models/res_users.py", line 22, in _check_session_validity
    delay, urls = ConfigParam.get_session_parameters()
  File "<string>", line 2, in get_session_parameters
  File "/opt/odoo/10/odoo/odoo/tools/cache.py", line 80, in lookup
    key = key0 + self.key(*args, **kwargs)
  File "<string>", line 1, in <lambda>
AttributeError: 'ir.config_parameter' object has no attribute 'cr'
```
I'm not familiar with the ormcache decorator, but this works as a workaround.
```
diff --git a/auth_session_timeout/models/ir_config_parameter.py b/auth_session_timeout/models/ir_config_parameter.py
index 83d8a64..8e2e87e 100644
--- a/auth_session_timeout/models/ir_config_parameter.py
+++ b/auth_session_timeout/models/ir_config_parameter.py
@@ -13,7 +13,7 @@ class IrConfigParameter(models.Model):
     _inherit = 'ir.config_parameter'
     @api.model
-    @tools.ormcache('self.cr.dbname')
+    @tools.ormcache('self.env.cr.dbname')
     def get_session_parameters(self):
         ConfigParam = self.env['ir.config_parameter']
         delay = ConfigParam.get_param(DELAY_KEY, 7200)
```

yajo commented:
Still ❌

pedrobaeza commented:
@chienandalu please review this one instead of your PR #875

chienandalu reviewed (commented):
@lasley Well, sorry for my part of contribution to this confusion...

Anyway, are you going to implement any of the changes made on https://github.com/OCA/server-tools/pull/887 or  https://github.com/OCA/server-tools/pull/875?

lasley commented:
> Anyway, are you going to implement any of the changes made on #887 or #875?

I'm going to have to go through the diffs to understand the changes. I'm hoping maybe a cherry pick of some stuff will do it 😆

jmorgannz committed:
Module auth_session_timeout: Pluggability (#887)

* Module auth_session_timeout:
---------------------------

* Refactor to allow other modules to inherit and augment or override the following:
** Session expiry time (deadline) calculation
** Ignored URLs
** Final session expiry (with possibility to late-abort)
* Re-ordered functionality to remove unnecessary work, as this code is called very often.
* Do not expire a session if delay gets set to zero (or unset / false)

* WIP

* Fixed flake8 lint errors

* Fixed flake8 lint errors

* WIP

* WIP

* WIP

* WIP

* WIP

* WIP

* Module: auth-session-timeout: Refactor ResUser tests to use `unittest.mock` patching

* Module: auth_session_timeout: Fixed flake8 lint errors

* Module: auth_session_timeout: Fixed flake8 lint errors

pedrobaeza commented:
@lasley can you finish this integrating #887 ?

lasley commented:
Cherry picked. My bad, I forgot this

pedrobaeza commented:
I'm getting this error installing this module:
```
Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/ir/ir_http.py", line 103, in _authenticate
    request.session.check_security()
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 1058, in check_security
    security.check(self.db, self.uid, self.password)
  File "/opt/odoo/custom/src/odoo/odoo/service/security.py", line 13, in check
    return res_users.check(db, uid, passwd)
TypeError: unbound method check() must be called with res.users instance as first argument (got unicode instance instead)
```

lasley commented:
Oops this also reminded me that some of these methods needed to be switched around now that we're not worrying about compat. I'll submit a fix in about 30m

lasley committed:
[IMP] auth_session_timeout: Deprecate backwards compat + improve
* Deprecate backwards compatibility methods that were retained during v9 rework
* Upgrade API and rename a few things for PEP-8
* Switch to HttpCase for tests
* Switch to isolated build

lasley commented:
Alright we're ready for a final review here, then I'll squash up

yajo commented:
Still no luck with tests 😕

lasley commented:
> Still no luck with tests 😕

Yeah got me dude. It seems we're fighting an uphill battle with Travis and I am damned close to  just removing tests from multiple PRs until 🍏

lasley commented:
Yup I've got nothing - it's failing on a bunch of modules unrelated to me. The option I see is to remove all tests from this module, but we obviously still have a timebomb on our hands.

From a global perspective, this is an issue I am encountering in numerous places. My immediate thought on a solution is that all modules should be tested independently from each other, instead of inline with each other. That covers up incompatibility problems though, so I'm at a loss.

hbrunn commented:
exclude/include is no solution?

lasley commented:
It is technically, but [we've already done this in `server-tools`](https://github.com/OCA/server-tools/blob/10.0/.travis.yml#L27), so we're just N+1ing IMO. It's also damned near impossible to find the failing/conflicting module without a bunch of hunt/pecking with includes and excludes. 

Biggest issue though is that only the first test case is available on Runbot, so we miss out on functional review of anything excluded.

lasley commented:
Welp exclude is what I went with - better than removing tests. Thanks for reminding me that exists Holger.

Hopefully our server-tools split helps with this. Or our MQT redesign. Or something.

Furthermore, we are finally 🍏  (+1y1m later 😆 )

yajo commented:
Yikes, It's not fun having tests that get broken when an addon is installed... Well, in any case could you squash the migration commits for merge?

lasley commented:
Squashed!
